subprojects/dependency-management/src/main/java/org/gradle/api/internal/catalog/AbstractProjectAccessorsSourceGenerator.java

```diff
@@ -60,24 +66,38 @@ protected void writeHeader(String packageName) throws IOException {
         writeLn();
     }
 
-    protected static String rootProjectName(ProjectDescriptor descriptor) {
+    protected static ProjectDescriptor rootProject(ProjectDescriptor descriptor) {
         ProjectDescriptor current = descriptor;
         while (current.getParent() != null) {
             current = current.getParent();
         }
-        return current.getName();
+        return current;
     }
 
     protected void writeProjectAccessor(String name, ProjectDescriptor descriptor) throws IOException {
+        ProjectDescriptor rootProject = rootProject(descriptor);
+        warnIfRootProjectNameNotSetExplicitly(rootProject);
+
         writeLn("    /**");
         String path = descriptor.getPath();
         writeLn("     * Creates a project dependency on the project at path \"" + path + "\"");
         writeLn("     */");
-        String returnType = toClassName(path, rootProjectName(descriptor));
+        String returnType = toClassName(path, rootProject.getName());
         writeLn("    public " +  returnType + " get" + name + "() { return new " + returnType + "(getFactory(), create(\"" + path + "\")); }");
         writeLn();
     }
 
+    private static void warnIfRootProjectNameNotSetExplicitly(ProjectDescriptor project) {
+        if (!(project instanceof DefaultProjectDescriptor)) {
+            return;
+        }
+        DefaultProjectDescriptor descriptor = (DefaultProjectDescriptor) project;
+        if (!descriptor.hasExplicitName()) {
+            LOGGER.warn("Root project name not explicitly set, so it is named based on its enclosing directory. This might impact the generated code and implicitly" +
```

Comment:
I would add to the message that this concerns project accessors. It is clear from the context, but not when the message shows up in logs.